examples/hardware_decoding/main.go

```diff
@@ -0,0 +1,190 @@
+	defer sw_f.Free()
+
+	// Alloc input format context
+	inputFormatContext := astiav.AllocFormatContext()
+	if inputFormatContext == nil {
+		log.Fatal(errors.New("main: input format context is nil"))
+	}
+	defer inputFormatContext.Free()
+
+	// Open input
+	if err := inputFormatContext.OpenInput(*input, nil, nil); err != nil {
+		log.Fatal(fmt.Errorf("main: opening input failed: %w", err))
+	}
+	defer inputFormatContext.CloseInput()
+
+	// Find stream info
+	if err := inputFormatContext.FindStreamInfo(nil); err != nil {
+		log.Fatal(fmt.Errorf("main: finding stream info failed: %w", err))
+	}
+
+	var hw_pix_fmt astiav.PixelFormat
+
+	// Loop through streams
+	streams := make(map[int]*stream) // Indexed by input stream index
+	for _, is := range inputFormatContext.Streams() {
+		var err error
+
+		// Only process video
+		if is.CodecParameters().MediaType() != astiav.MediaTypeVideo {
+			continue
+		}
+
+		// Create stream
+		s := &stream{inputStream: is}
+
+		// Find decoder
+		if s.decCodec = astiav.FindDecoder(is.CodecParameters().CodecID()); s.decCodec == nil {
+			log.Fatal(errors.New("main: codec is nil"))
+		}
+
+		// Alloc codec context
+		if s.decCodecContext = astiav.AllocCodecContext(s.decCodec); s.decCodecContext == nil {
+			log.Fatal(errors.New("main: codec context is nil"))
+		}
+		defer s.decCodecContext.Free()
+
+		hw_pix_fmt, err = astiav.FindSuitableHardwareFormat(s.decCodec, hw_device)
+		if err != nil {
+			log.Fatal(fmt.Errorf("main: find decoder hw format fails: %w", err))
+		} else {
+			log.Printf("Using hw_pix_fmt: %s", hw_pix_fmt.Name())
+		}
+
+		// Update codec context
+		if err := is.CodecParameters().ToCodecContext(s.decCodecContext); err != nil {
+			log.Fatal(fmt.Errorf("main: updating codec context failed: %w", err))
+		}
+
+		if s.hwDeviceContext, err = hw_decoder_init(s.decCodecContext, hw_device); err != nil {
+			log.Fatal(fmt.Errorf("main: init hardware device failed: %w", err))
+		}
+
+		// Open codec context
+		if err := s.decCodecContext.Open(s.decCodec, nil); err != nil {
+			log.Fatal(fmt.Errorf("main: opening codec context failed: %w", err))
+		}
+
+		// Add stream
+		streams[is.Index()] = s
+	}
+
+	// Loop through packets
+	for {
+		// Read frame
+		if err := inputFormatContext.ReadFrame(pkt); err != nil {
+			if errors.Is(err, astiav.ErrEof) {
+				break
+			}
+			log.Fatal(fmt.Errorf("main: reading frame failed: %w", err))
+		}
+
+		// Get stream
+		s, ok := streams[pkt.StreamIndex()]
+		if !ok {
+			continue
+		}
+
+		// Send packet
+		if err := s.decCodecContext.SendPacket(pkt); err != nil {
+			log.Fatal(fmt.Errorf("main: sending packet failed: %w", err))
+		}
+
+		// Loop
+		for {
+			// Receive frame
+			if err := s.decCodecContext.ReceiveFrame(gpu_f); err != nil {
+				if errors.Is(err, astiav.ErrEof) || errors.Is(err, astiav.ErrEagain) {
+					break
+				}
+				log.Fatal(fmt.Errorf("main: receiving frame failed: %w", err))
+			}
+
+			if gpu_f.PixelFormat() == hw_pix_fmt {
```

Comment:
you sad i should rename gpu_f to hardwareFrame and sw_f to softwareFrame what would now be the finalFrame ? 
Now it would look like this.


```go
// Loop
for {
	// Receive frame
	if err := s.decCodecContext.ReceiveFrame(hardwareFrame); err != nil {
		if errors.Is(err, astiav.ErrEof) || errors.Is(err, astiav.ErrEagain) {
			break
		}
		log.Fatal(fmt.Errorf("main: receiving frame failed: %w", err))
	}

	// Get frame
	if hardwareFrame.PixelFormat() == hardwarePixelFormat {

		// Transfer hardware data
		if err := hardwareFrame.TransferHardwareData(softwareFrame); err != nil {
			log.Fatal(fmt.Errorf("main: Unable to transfer frame from gpu: %w", err))
		}

		softwareFrame.SetPts(hardwareFrame.Pts())
	}

	log.Printf("new frame: stream %d - pts: %d - data transferred from hardware: %v", pkt.StreamIndex(), softwareFrame.Pts(), hardwareFrame.PixelFormat() == hardwarePixelFormat)
}
```

Reply:
That way would be good:
```go
// Loop
for {
	// Receive frame
	if err := s.decCodecContext.ReceiveFrame(hardwareFrame); err != nil {
		if errors.Is(err, astiav.ErrEof) || errors.Is(err, astiav.ErrEagain) {
			break
		}
		log.Printf("main: receiving frame failed: %s", err.Error())
	}

	// Get frame
        var finalFrame *astiav.Frame
	if hardwareFrame.PixelFormat() == hardwarePixelFormat {
		// Transfer hardware data
		if err := hardwareFrame.TransferHardwareData(softwareFrame); err != nil {
			log.Fatal(fmt.Errorf("main: Unable to transfer frame from gpu: %w", err))
		}

                // Update pts
		softwareFrame.SetPts(hardwareFrame.Pts())

                // Update final frame
                finalFrame = softwareFrame
	} else {
                // Update final frame
                finalFrame = hardwareFrame
        }

        // Do something with decoded frame
	log.Printf("new frame: stream %d - pts: %d - data transferred from hardware: %v", pkt.StreamIndex(), finalFrame.Pts(), finalFrame.PixelFormat() == hardwarePixelFormat)
}
```

Reply:
ok make sense =)

Reply:
👍 

By the way, `log.Printf("main: receiving frame failed: %s", err.Error())` should be `log.Fatal(fmt.Errorf("main: receiving frame failed: %w", err))` 👍

Reply:
thx